[wip] Clickhouse exporter improvements for logs

mauidude opened the pull request:
WIP DO NOT MERGE YET

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/ClickHouse/opentelemetry-collector-contrib?pullRequest=1) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/ClickHouse/opentelemetry-collector-contrib?pullRequest=1) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/ClickHouse/opentelemetry-collector-contrib?pullRequest=1) it.</sub>

mauidude commented on Makefile:
```diff
@@ -221,7 +221,7 @@ run:
 	GOOS=linux GOARCH=amd64 $(MAKE) $(COMPONENT)
 	cp ./bin/$(COMPONENT)_linux_amd64 ./cmd/$(COMPONENT)/$(COMPONENT)
-	docker build -t $(COMPONENT) ./cmd/$(COMPONENT)/
+	docker buildx build --platform=linux/amd64 -t $(COMPONENT) ./cmd/$(COMPONENT)/
```
wasn't able to build on m1 mac without this

mauidude commented on cmd/otelcontribcol/Dockerfile:
```diff
@@ -9,7 +9,8 @@ ARG USER_UID=10001
 
 COPY --from=prep /etc/ssl/certs/ca-certificates.crt /etc/ssl/certs/ca-certificates.crt
-COPY otelcontribcol /
+COPY otelcontribcol /otelcol-contrib
```
their Dockerfiles are confusing... the helm chart was expecting to run the command `/otelcol-contrib` but the binary that gets built and copied into the image is `/otelcontribcol`?

mauidude commented on exporter/clickhouseexporter/exporter_logs.go:
```diff
@@ -73,28 +73,40 @@ func (e *logsExporter) shutdown(_ context.Context) error {
-	err := doWithTx(ctx, e.client, func(tx *sql.Tx) error {
-		statement, err := tx.PrepareContext(ctx, e.insertSQL)
+	err := func() error {
+		scope, err := e.client.Begin()
```
this is apparently the way to do batch inserts using the database/sql interface of the clickhouse driver https://clickhouse.com/docs/en/integrations/go/clickhouse-go/database-sql-api#batch-insert

mauidude commented on exporter/clickhouseexporter/exporter_logs.go:
```diff
@@ -238,17 +246,3 @@ func renderCreateLogsTableSQL(cfg *Config) string {
 	return fmt.Sprintf(insertLogsSQLTemplate, cfg.LogsTableName)
 }
-
-func doWithTx(_ context.Context, db *sql.DB, fn func(tx *sql.Tx) error) error {
```
this was not the proper way, they were using a tx but they were using an insert per row which is not correct, it's one  prepared insert and then you append each row to the batch using `Exec()`

vadimp commented:
Actually it doesn't look like we've changed anything, but reduced the consistency of the insertion.
Underneath,  [Prepare](https://github.com/golang/go/blob/70f98a251efdbfd619c4ff466a43da299ad04752/src/database/sql/sql.go#L2352) calls [PrepareContext](https://github.com/golang/go/blob/70f98a251efdbfd619c4ff466a43da299ad04752/src/database/sql/sql.go#L1556), and [Exec](https://github.com/golang/go/blob/70f98a251efdbfd619c4ff466a43da299ad04752/src/database/sql/sql.go#L2622) calls [ExecContext](https://github.com/golang/go/blob/70f98a251efdbfd619c4ff466a43da299ad04752/src/database/sql/sql.go#L2598).

Assuming I've not messed up with sql wrapper interfaces and followed the code correctly, the only difference is removal of transaction semantics that have nothing more than some consistency guarantees in writing data across partitions. Doesn't mean is a bad thing for our case.

Both `Exec` and `ExecContext` execute the query immediately and return an empty result set in our case. So the roundtrip executed in both cases.

```go
func (s *Stmt) ExecContext(ctx context.Context, args ...any) (Result, error) {
	s.closemu.RLock()
	defer s.closemu.RUnlock()

	var res Result
	err := s.db.retry(func(strategy connReuseStrategy) error {
		dc, releaseConn, ds, err := s.connStmt(ctx, strategy)
		if err != nil {
			return err
		}

		res, err = resultFromStatement(ctx, dc.ci, ds, args...)
		releaseConn(err)
		return err
	})

	return res, err
}
```

The real performance improvement will come from bulk inserts based on Append
Example here:
https://github.com/ClickHouse/clickhouse-go/blob/main/examples/clickhouse_api/batch.go

My assumption is that the performance improvement we see is coming from the usage of gateway that results in lower parallelism and as a result more "coordination" in comparison to much higher number of collectors running as DaemonSets.

mauidude commented:
According to their [docs](https://pkg.go.dev/github.com/ClickHouse/clickhouse-go/v2#section-readme) and [examples](https://github.com/ClickHouse/clickhouse-go/blob/main/examples/clickhouse_api/batch.go), this is how you do batching with the golang `database/sql` api. The changes I made, were to do this **correctly**. The way it was previously implemented, was doing each individual log as individually prepared `INSERT INTO table (...) VALUES (...)` statements (for each log -> open tx, prepare insert, exec insert), when you are supposed to use a prepared statement for the entire batch within a transaction (open tx, prepare, for each row -> append log, commit). It was not batching them previously.

> Actually it doesn't look like we've changed anything, but reduced the consistency of the insertion.

I have to verify but I believe we've actually increased the consistency as I believe the batch is atomic. I am not convinced that the transactions of the previous implementation did anything because there aren't really transactional semantics in ClickHouse. However, if the batched insert meets [these conditions](https://github.com/ClickHouse/ClickHouse/issues/9195#issuecomment-587500824) then it is my understanding that it would be "atomic". I believe our inserts satisfy these critieria or the configs can be tuned to make them satisfied (such as increasing the max_insert_block_size if needed).

> The real performance improvement will come from bulk inserts based on Append

This is what I am doing but using the database/sql api instead of the clickhouse driver api directly.  I had implemented it this way originally, but realized they were the same thing and then decided to keep the sql interface since the initialization of the driver was a bit different and I was worried about having to add more configurations. The database/sql api seemed easier to integrate since it was already being used.

> My assumption is that the performance improvement we see is coming from the usage of gateway

Correct, this was already in place when I introduced my changes to the exporter, which seemed to have fixed reduced the memory usage of the agents, and also fixed the insertion issues (large number of parts, cluster instability, etc...) into clickhouse. It doesn't mean that these changes aren't necessary as there are clear misunderstandings of how to insert into clickhouse in the code base. I also only changed the logging code. Metrics and traces have similar issues that need to be updated but I will probably defer those until later.

vadimp commented:
> I have to verify

I looked into similar database/std [batch example ](https://github.com/ClickHouse/clickhouse-go/blob/main/examples/std/batch.go) to verify it by debugging the flow. This example is the same flow we run. Indeed, the underlying implementation batches "values" in memory before the execution. 

<img width="503" alt="image" src="https://user-images.githubusercontent.com/657528/224989196-0e1eab53-bbd6-4426-9657-b9bd86e1e4d9.png">


<img width="760" alt="image" src="https://user-images.githubusercontent.com/657528/224988268-06445e14-517d-48ee-af01-36d37d5bf55e.png">

mauidude commented:
i updated the code to use the native protocol, which isn't supported by the database/sql interface
